cps/main.c

```diff
@@ -16,11 +16,529 @@
+
+	memset(&ops, 0, sizeof(ops));
+	ops.port_id = conf.group_id;
+	ops.change_mtu = kni_change_mtu;
+	ops.config_network_if = kni_change_if;
+
+	*kni = rte_kni_alloc(mp, &conf, &ops);
+	if (*kni == NULL) {
+		RTE_LOG(ERR, KNI, "Could not allocate KNI for %s iface\n",
+			iface->name);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+cps_stage1(void *arg)
+{
+	struct cps_config *cps_conf = arg;
+	unsigned int socket_id = rte_lcore_to_socket_id(cps_conf->lcore_id);
+	int ret;
+
+	ret = assign_cps_queue_ids(cps_conf);
+	if (ret < 0)
+		goto error;
+
+	ret = kni_create(&cps_conf->front_kni,
+		cps_conf->net->gatekeeper_pktmbuf_pool[socket_id],
+		&cps_conf->net->front);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: failed to create KNI for the front iface\n");
+		goto error;
+	}
+
+	if (cps_conf->net->back_iface_enabled) {
+		ret = kni_create(&cps_conf->back_kni,
+			cps_conf->net->gatekeeper_pktmbuf_pool[socket_id],
+			&cps_conf->net->back);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"cps: failed to create KNI for the back iface\n");
+			goto error;
+		}
+	}
+
+	return 0;
+
+error:
+	cleanup_cps();
+	return ret;
+}
+
+static void
+fill_bgp_rule(struct ipv6_acl_rule *rule, struct gatekeeper_if *iface,
+	int filter_source_port)
+{
+	uint32_t *ptr32 = (uint32_t *)&iface->ip6_addr.s6_addr;
+	int i;
+
+	rule->data.category_mask = 0x1;
+	rule->data.priority = 1;
+	/* Userdata is filled in in register_ipv6_acl(). */
+
+	rule->field[PROTO_FIELD_IPV6].value.u8 = IPPROTO_TCP;
+	rule->field[PROTO_FIELD_IPV6].mask_range.u8 = 0xFF;
+
+	for (i = DST1_FIELD_IPV6; i <= DST4_FIELD_IPV6; i++) {
+		rule->field[i].value.u32 = rte_be_to_cpu_32(*ptr32);
+		rule->field[i].mask_range.u32 = 32;
+		ptr32++;
+	}
+
+	if (filter_source_port) {
+		rule->field[SRCP_FIELD_IPV6].value.u16 = TCP_PORT_BGP;
+		rule->field[SRCP_FIELD_IPV6].mask_range.u16 = 0xFFFF;
+	} else {
+		rule->field[DSTP_FIELD_IPV6].value.u16 = TCP_PORT_BGP;
+		rule->field[DSTP_FIELD_IPV6].mask_range.u16 = 0xFFFF;
+	}
+}
+
+static int
+add_bgp_filters(struct gatekeeper_if *iface, uint16_t rx_queue)
+{
+	if (ipv4_if_configured(iface)) {
+		int ret;
+		ret = ntuple_filter_add(iface->id, iface->ip4_addr.s_addr,
+			rte_cpu_to_be_16(TCP_PORT_BGP), UINT16_MAX, 0, 0,
+			IPPROTO_TCP, rx_queue, true);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"cps: could not add source BGP filter on %s iface\n",
+				iface->name);
+			return ret;
+		}
+		ret = ntuple_filter_add(iface->id, iface->ip4_addr.s_addr,
+			0, 0, rte_cpu_to_be_16(TCP_PORT_BGP), UINT16_MAX,
+			IPPROTO_TCP, rx_queue, true);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"cps: could not add destination BGP filter on %s iface\n",
+				iface->name);
+			return ret;
+		}
+	}
+
+	if (ipv6_if_configured(iface)) {
+		struct ipv6_acl_rule ipv6_rules[NUM_ACL_BGP_RULES];
+		int ret;
+
+		memset(&ipv6_rules, 0, sizeof(ipv6_rules));
+
+		fill_bgp_rule(&ipv6_rules[0], iface, true);
+		fill_bgp_rule(&ipv6_rules[1], iface, false);
```

Comment:
Add the following comment for the first filter:
`/* Capture packets for connections that our BGP speaker started. */`

And the following comment for the second filter:
`/* Capture packets for connections that remote BGP speakers started. */`